File: leetcode/q106/Recursion1.java
package leetcode.q106;

import java.util.Map;
import java.util.HashMap;

public class Recursion1 {
    
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        Map<Integer, Integer> inMap = new HashMap<>();
        for (int i = 0; i < inorder.length; i++) {
            inMap.put(inorder[i], i);
        }
        
        return buildTree(postorder, 0, postorder.length - 1, inMap, 0, inorder.length - 1);
    }
    
    private TreeNode buildTree(int[] post, int pLeft, int pRight, Map<Integer, Integer> inMap, int iLeft, int iRight) {
        if (pLeft > pRight) {
            return null;
        }
        
        TreeNode root = new TreeNode(post[pRight]);
        int iIndex = inMap.get(post[pRight]);
        int rightSize = iRight - iIndex;
        root.left = buildTree(post, pLeft, pRight - rightSize - 1, inMap, iLeft, iIndex - 1);
        root.right = buildTree(post, pRight - rightSize, pRight - 1, inMap, iIndex + 1, iRight);
        return root;
    }
    
    static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int val) {
            this.val = val;
            left = right = null;
        }
    }
}

File: leetcode/q106/recursion1.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;


class Solution {
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        unordered_map<int, int> inmap;
        for (int i = 0; i < inorder.size(); i++) {
            inmap[inorder[i]] = i;
        }
        return buildTree(postorder, 0, postorder.size() - 1, inmap, 0, inorder.size() - 1);
    }
private:
    TreeNode* buildTree(const vector<int>& postorder, int pLeft, int pRight, const unordered_map<int, int>& inmap, int inLeft, int inRight) {
        if (pLeft > pRight) {
            return NULL;
        }
        int val = postorder[pRight];
        TreeNode* root = new TreeNode(val);
        int inIndex = inmap.at(val);
        int rightSize = inRight - inIndex;
        root->left = buildTree(postorder, pLeft, pRight - rightSize - 1, inmap, inLeft, inIndex - 1);
        root->right = buildTree(postorder, pRight - rightSize, pRight - 1, inmap, inIndex + 1, inRight);
        return root;
    }
};

class TreeNode {
public:
    int val;
    TreeNode* left;
    TreeNode* right;

    TreeNode() : val(val), left(NULL), right(NULL) {}
    TreeNode(int val) : left(NULL), right(NULL) {
        this->val = val;
    }
};
